File: src/core/helm/base/api/os/OperatingSystem.java
/*
 * Copyright (C) 2023 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.fullstack.base.api.os;

public enum OperatingSystem {
    WINDOWS("windows"),
    LINUX("linux"),
    DARWIN("darwin");

    private final String directoryName;

    OperatingSystem(final String directoryName) {
        this.directoryName = directoryName;
    }

    public String directoryName() {
        return directoryName;
    }

    /**
     * Attempts to determine the current operating system based on the system property "os.name".
     *
     * @return the current operating system.
     * @throws IllegalStateException if the current operating system is not supported.
     */
    public static OperatingSystem current() {
        final String osName = System.getProperty("os.name").toLowerCase();
        if (osName.contains("win")) {
            return WINDOWS;
        } else if (osName.contains("nix") || osName.contains("nux") || osName.contains("aix")) {
            return LINUX;
        } else if (osName.contains("mac")) {
            return DARWIN;
        } else {
            throw new IllegalStateException("Unsupported operating system: " + osName);
        }
    }
}

File: src/core/helm/base/api/os/OperatingSystem.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Represents supported operating systems.
 */
export enum OperatingSystem {
  WINDOWS = 'windows',
  LINUX = 'linux',
  DARWIN = 'darwin',
}

export namespace OperatingSystem {
  /**
   * Returns the directory name associated with an operating system.
   * @param os - The operating system enum value
   * @returns The directory name for the operating system
   */
  export function directoryName(os: OperatingSystem): string {
    return os;
  }

  /**
   * Attempts to determine the current operating system based on process.platform.
   *
   * @returns The current operating system.
   * @throws Error if the current operating system is not supported.
   */
  export function current(): OperatingSystem {
    const platform = process.platform;

    if (platform === 'win32') {
      return OperatingSystem.WINDOWS;
    } else if (platform === 'linux') {
      return OperatingSystem.LINUX;
    } else if (platform === 'darwin') {
      return OperatingSystem.DARWIN;
    }

    throw new Error(`Unsupported operating system: ${platform}`);
  }
}
